GitCommands/Git/Commands/GitCommandHelpers.cs

```diff
@@ -81,39 +80,54 @@ public static ArgumentString AddSubmoduleCmd(string remotePath, string localPath
-                        : default)
-                {
-                    { branches, "--dereference", "--tags" },
-                };
-            }
-            else if (branches)
-            {
-                // branches only
-                cmd = new GitArgumentBuilder("for-each-ref", gitOptions:
-                    noLocks && GitVersion.Current.SupportNoOptionalLocks
-                        ? (ArgumentString)"--no-optional-locks"
-                        : default)
-                {
-                    "--sort=-committerdate",
-                    @"refs/heads/",
-                    @"--format=""%(objectname) %(refname)"""
-                };
+                // If we don't need tags, it is easy.
+                format = @"--format=""%(objectname) %(refname)""";
             }
             else
             {
-                throw new ArgumentException("GetRefs: Neither branches nor tags requested");
+                // ...however, if we're interested in tags, tags may be simple (in which case they are point to commits directly),
+                // or "dereferences" (i.e. commits that contian metadata and point to other commits, "^{}").
+                // Derefence commits do not contain date information, so we need to find information from the referenced commits (those with '*').
+                // So the following format is as follows:
+                //      If (there is a 'authordate' information, then this is a simple tag/direct commit)
+                //      Then
+                //          format = %(objectname) %(refname)
+                //      Else
+                //          format = %(*objectname) %(*refname) // i.e. info from a referenced commit
+                //      End
+                format = @"--format=""%(if)%(authordate)%(then)%(objectname) %(refname)%(else)%(*objectname) %(*refname)%(end)""";
             }
 
+            GitArgumentBuilder cmd = new GitArgumentBuilder("for-each-ref",
+                gitOptions: noLocks && GitVersion.Current.SupportNoOptionalLocks
+                    ? (ArgumentString)"--no-optional-locks"
+                    : default)
+            {
+                { sortBy != GitRefsSortBy.Default, GetSortCriteria(tags, sortBy, sortOrder), string.Empty },
+                format,
+                { branches, "refs/heads/", string.Empty },
+                { branches && tags, "refs/remotes/", string.Empty },
+                { tags, "refs/tags/", string.Empty },
+            };
+
             return cmd;
+
+            static string GetSortCriteria(bool needTags, GitRefsSortBy sortBy, GitRefsSortOrder sortOrder)
+            {
+                string order = sortOrder == GitRefsSortOrder.Ascending ? string.Empty : "-";
+                if (!needTags)
+                {
+                    return $"--sort={order}{sortBy}";
+                }
+
+                // Sort by dereferenced data
+                return $"--sort={order}{sortBy} --sort={order}*{sortBy}";
```

Comment:
Add a comment that Git 2.14.6 is required (maybe remove checks and add comments elsewhere?)
There is a workaround, to not set any sorting too
https://github.com/gitextensions/gitextensions/pull/8427#discussion_r483852667

Reply:
Added `GitVersion.Current.SupportRefSort`